src/toxcore/dht_new/dht_impl.rs

```diff
@@ -0,0 +1,297 @@
+    {
+        debug!(target: "DhtPacket", "Getting packet data from DhtPacket.");
+        trace!(target: "DhtPacket", "With DhtPacket: {:?}", self);
+        let decrypted = open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
+            .map_err(|e| {
+                debug!("Decrypting DhtPacket failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtPacket decrypt error: {:?}", e))
+            });
+
+        match DhtPacketPayload::from_bytes(&decrypted.unwrap_or(vec![0]), self.packet_kind) {
+            IResult::Incomplete(e) => {
+                error!(target: "DhtPacket", "PingRequest deserialize error: {:?}", e);
+                Err(Error::new(ErrorKind::Other,
+                    format!("PingRequest deserialize error: {:?}", e)))
+            },
+            IResult::Error(e) => {
+                error!(target: "DhtPacket", "PingRequest deserialize error: {:?}", e);
+                Err(Error::new(ErrorKind::Other,
+                    format!("PingRequest deserialize error: {:?}", e)))
+            },
+            IResult::Done(_, packet) => {
+                Ok(Some(packet))
+            }
+        }
+    }
+}
+
+impl DhtRequest {
+    /// create new DhtRequest object
+    pub fn new(shared_secret: &PrecomputedKey, rpk: &PublicKey, spk: &PublicKey, dp: DhtRequestPayload) -> DhtRequest {
+        let nonce = &gen_nonce();
+
+        let mut buf = [0; MAX_DHT_PACKET_SIZE];
+        let (_, size) = dp.to_bytes((&mut buf, 0)).unwrap();
+        let payload = seal_precomputed(&buf[..size], nonce, shared_secret);
+
+        DhtRequest {
+            rpk: *rpk,
+            spk: *spk,
+            nonce: *nonce,
+            payload: payload,
+        }
+    }
+
+    /**
+    Decrypt payload and try to parse it as packet type.
+
+    Unlike DhtPacket, DhtRequest has its type in encrypted payload.
+    So, first decrypt and then check its packet kind.
+
+    Returns `Error` in case of faliure:
+
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtRequestPayload>, Error>
+    {
+        debug!(target: "DhtRequest", "Getting packet data from DhtRequest.");
+        trace!(target: "DhtRequest", "With DhtRequest: {:?}", self);
+        let decrypted = open(&self.payload, &self.nonce, &self.spk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
+            .map_err(|e| {
+                debug!("Decrypting DhtRequest failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtRequest decrypt error: {:?}", e))
+            });
+
+        // size    | description
+        // 1       | nat ping req|resp (0xfe)
+        // 1       | Request or Response flag
+        // 8       | Request Id (Ping Id)
+        let packet_payload = decrypted.unwrap_or(vec![0,0]);
```

Comment:
I'd propose to do here the same thing as for DhtPacket::get_payload to reduce code size.